### ISSUE TITLE
feat(animation): Allow binding expression in style(), keyframes() etc...

### ISSUE THREAD

**Issue by shlomiassaf** [angular/angular#9668]
**I'm submitting a ...**  (check one with "x")
feature request

**Current behavior**
Animation instructions doesn't accept bindings expression.

**Expected/desired behavior**
Animation metadata should allow instructions from the component instance.
 
Example: 
```ts
@Component({
  selector: 'my-app',
  animations: [
    trigger('fade', [
      transition('void => in', [
        animate('150ms linear', keyframes([
          style('MY_STYLE'),
          style({opacity: '*'})
        ]))
      ])
    ])
  ],
  encapsulation: ViewEncapsulation.None,
  template: `<div  @fade="fadeState"></div>`})
export class FadeInBackdrop {
  private fadeState: 'in' | 'out' = 'in';
  MY_STATE = {
    opacity: 0
  }
}
```

**What is the motivation / use case for changing the behavior?**
This will allow dynamic animation based on user input.
Sine animation instructions are static the only way to change them in runtime is
A) Using auto style, letting the user set * from css
B) Define multiple states (not really dynamic)

This is really useful in 3rd part libraries where some sort of animation customization is needed.

**Comment by matsko** [comment]
We have an API in mind for this actually:

```ts
@Component({
  animations: [
    trigger('fade', [
       style('$myStyle')
    ])
  ],
  template: `
    <div @fade="fadeState"></div>
  `
})
export class FadeInBackdrop {
  public currentFadeState = 'in';
  private fadeStateValues = {
     in: {
       value: 'in',
       data: { 'myStyle': { opacity: 1 } }
     },
     out: {
       value: 'out',
       data: { 'myStyle': { opacity: 0 } }
     }
  }

  get fadeState() {
    return this.fadeStateValues[this.currentFadeState];
  }
}
```

**Comment by shlomiassaf** [comment]
👍 

Took me some time to get this 

why the `value` property in the state interface? I don't see it used.

**Comment by matsko** [comment]
It's the value of the state:

```
<div @fadeState="stateVal"></div>

// string-based value
stateVal = 'in';

// is the same as
stateVal = { value: 'in', data { ... }}

```

**Comment by shlomiassaf** [comment]
I can see angular 3rd party animation implementation libraries coming after this lands :)

**Comment by matsko** [comment]
3rd party tools will use this API:

```ts
animations: [
   animation('3rdPartyAnimation', [
      style('$value'),
      animate('$duration', { prop: '$value2' })
   ]),

   trigger('something', [
      transition('* => *', animation('3rdPartyAnimation'))
   ])
]
```

**Comment by dpix** [comment]
Any Idea on which release candidate this will make it into? I've just upgraded to RC4 to check out animations but I still need to calculate values for an animation at run-time. Unless there is some way to do this currently?

eg:
```javascript
 animations: [
        trigger('slide', [
            state('0', style({'margin-left': someCalculatedValueOrFunction })),

```

**Comment by matsko** [comment]
Yes we do have an API in mind for this. But getting some of the core bugs out of the way for final is a higher priority. Once the CSS Parser integration and query support are in then this will get in (hopefully in the coming weeks).

The API for this will be: 

```ts
// register in bootstrap
function myCalculationFn(element, styles, prop, args) {
   styles[prop] = ...
}

bootstrap(App, [
   provide(ANIMATION_HELPERS, { multi: true, useValue: myCalculationFn })
])

// in your component
animations: [
        trigger('slide', [
            state('0', style({'margin-left': 'myCalculateFn()' })),
```

**Comment by dharders** [comment]
The new API looks very useful @matsko . Quick question: Does the Angular (or web animation) Framework enable a snapshot of current styles applied at the time of state change ? i.e. So we can pass that value to myCalculateFn() and animate back to the state value from the current in flight animation to avoid jarring (i.e. Greensock GSAP TweenMax.to() functionality). Related issue #9661

**Comment by matsko** [comment]
Yes there is something in mind for an `:initial` or `:start` CSS object. The API is still being figured out.

**Comment by dharders** [comment]
Great. Looking forward to it!

**Comment by helzgate** [comment]
RC7 was just released.  Is this something we can test in RC7 yet?

**Comment by vladtar** [comment]
Any updates on this?

**Comment by quickbacon** [comment]
I am looking forward to having runtime control over animation styles. Currently I have to rely on jquery for these more custom animations and it's messy. Will this be available soon?

**Comment by mattheiler** [comment]
+1. Is there a status on this?

**Comment by andrewheusser** [comment]
+1!

**Comment by Fredsters** [comment]
Are there any updates?

**Comment by lazarljubenovic** [comment]
Just let the guys work already. It's not like they will do a feature and then lock in it their basement until we bore them to death. Just subscribe to the thread and you will receive notifications when it's done, or when the team has updates.

**Comment by borislemke** [comment]
Hey guys, it's been almost a year now and I stumbled across this thread. Is this still open? Any reference on how to use this if it's been implemented?

**Comment by vaibhav93** [comment]
Any updates ?

**Comment by m-abs** [comment]
Our project could really use this.

We've had to use transitions in our CSS together with `setTimeout` to do our animations.

**Comment by realappie** [comment]
@matsko Any updates on this thread?

**Comment by tomwanzek** [comment]
First of all, thanks for all the hard work on Angular!

A couple of comments:

(1) 

As far as animations in Angular are concerned, the support of a pattern allowing transitions based on changes in bound data is absolutely paramount. Support should cover both discrete and continuous state spaces.

Not having this support unfortunately renders Angular mute for the kind of data visualization animations I mostly require. With proper support, I could have a clean separation of concerns when using Angular and D3:

* D3 for pure layout computations (no DOM manipulation, no binding data to DOM elements)
* Angular for any and all DOM interaction.

Without a supported pattern, I need to flow the animations through D3, meaning:
* importing otherwise unnecessary D3 modules (d3-selection, d3-transition)
* Having D3 select DOM elements, bind transition related data to them and manage the transitions through D3 as part of the Angular life cycle hooks.

(2)

When experimenting with Angular animations (including using `'*'` for auto calc), I noticed that `<svg>` element attributes, where supported according to Web Animations API, are not working with Angular animations. 

E.g. animating an `<svg:circle>` radius attribute `r` works only on Chrome (where the `style` setter seems to serve as a "fallback" for an `attribute` setter) But since Firefox and Edge seem to cleanly separate css-style setting from svg-attribute setting, the Angular animations implementation does not seem to extend to `<svg>` animatable attributes.

The main reason, I mention (2) here, is that I would hope/expect that Angular animations with keybindings also address this scenario. Otherwise, I seems I should open a separate issue for it.

**Comment by citizentim** [comment]
Pace @lazarljubenovic, as a product manager myself, the number of people asking for a particular feature is a critical factor in determining how to prioritise said feature.  "Shut up" isn't a very good way to help product managers get the data they need to make good decisions on prioritisation.

Keep the comments coming, and keep them respectful, that's all.  And I think everyone has been respectful, so there's no issue.

I am therefore asking that this RFE be prioritised for release in the short-term.  The lack of ability to bind animation values is a major limitation of Angular animation, and perhaps I'm wrong, but it doesn't seem very expensive in terms of dev cycles.  I know that I can't do my project properly without it.

Respectfully.

**Comment by glinkot** [comment]
I'm trying to integrate GSAP and not having much luck, this support would be really valuable.

**Comment by OSUblake** [comment]
@glinkot Here's an example of how you can do state based animations with GSAP. It's not perfect, but it gives you a lot more control over your animations.

https://www.webpackbin.com/bins/-KlC3WIhC60G0V7Ck5CN

**Comment by glinkot** [comment]
Very cool, thanks Blake!  Hadn't seen Webpackbin, I'd been looking for a good spot to sandbox Angular things as codepen doesn't make it easy!  Hopefully it allows usage of all the GSAP stuff like codepen does, but that's not the topic of the thread so I'll ask that somewhere else I guess ;)

**Comment by realappie** [comment]
For any of you who are curious how the new API looks like, have a look at [this](https://youtu.be/Oh9wj-1p2BM) video

**Comment by manuelnaturalcycles** [comment]
even with the new api and following [this article](https://www.yearofmoo.com/2017/06/new-wave-of-animation-features.html#what-about-binding-level-options) written by the author, I still cannot make this work.
-- animations.util.ts
`export const fadeAnimation =
  animation([
    style({
      opacity: "{{ from }}"
    }),
    animate("{{ time }}",
      style({
        opacity: "{{ to }}"
      })
    )
  ], {
    params: {
      time: "1s",
      from: 0,
      to: 1
    }
  })`

-- component.ts
`@Component({
 //....
  animations: [
    trigger('fadeAnimation', [
      transition('* => *', [
        useAnimation(fadeAnimation)
      ])
    ])
  ]
})`

--component.html
`<div class="container registerCredentials" [@fadeAnimation]="{time: '20s', from: 0, to: 1 }">`


the animation works but it uses time as 1s instead of 20s... i just want to know how to use data from the controller as input of the params of the animation, i wonder if the new version supports this

**Comment by matsko** [comment]
Try:

```html
<div [@fadeAnimation]="{value: 'something', params: { time: '20s', from: 0, to: 1 }}">
```

**Comment by manuelnaturalcycles** [comment]
amazing! thanks!

**Comment by tomwanzek** [comment]
@matsko Thanks for all the hard work on the animations features 💯 . A couple of questions related to your answer above:
1. Is it possible to update your wonderful [blogpost](https://www.yearofmoo.com/2017/06/new-wave-of-animation-features.html#what-about-binding-level-options) to use the current API which wraps the inputs into `params: {...}`? Similar to @manuelnaturalcycles I had found the `params` property used in the API documentation for `animate(...)`, `trigger(...)` etc, but failed to see it for the data binding with `[@xyz]=...`.
2. What is the correct way to persist the end state of an animation driven by input parameters? While I get the animation with input parameters working as per your comment above, at its end, it  does not persist. `state(...)` does not seem to support input parameters.

Thanks in advance!

**Comment by manuelnaturalcycles** [comment]
@tomwanzek i have that same problem in which the animation executes properly but the end state is not persisted. so if the element has no opacity value defined in css -> the animation will take it from 0 to 1 -> and then it goes back to undefined as initial state.

**Comment by tomwanzek** [comment]
@manuelnaturalcycles As an initial step, I am playing with using `[style.opacity]="fadeStatus.params.end"` on the DOM element being animated. It achieves the persistence at first glance while still allowing the animation to run.

As a side note: If I build the animation manually with `AnimationBuilder`, the end style of the animation is persisted without extra steps.

In the end, the catch for me is that I have to now test out my entire use case with the new features. This amounts to testing out dynamic animations between old and new input values to a component with `onPush` change detection and corresponding lifecycle hooks. My underlying state space is continuous, so I may have to map it through a synthetic discrete state space just to trigger the transitions. I will comment back here, when I had a chance to "go full cycle"... Hopefully successful.

**Comment by tomwanzek** [comment]
@manuelnaturalcycles I have created a repo with an initial few examples on the data-driven animations I had in mind: [ngx-anime](https://github.com/tomwanzek/ngx-anime). A live demo is available [here](https://tomwanzek.github.io/ngx-anime/).

Based on the approach I took so far (and there might be other/better ones). Here are a couple of early observations:
* Given how the animation triggers currently work (i.p. the bindable state object with `params` since 4.2), data-driven animations off component inputs are now possible in principle ( 💃 )
* Persistence can be achieved by binding the "end" state value to the relevant style property of the animated element using `[style.xyz]=abcState.params.end` (This is in absence of input parameters in the trigger's `state(...)` configuration, see my question to @matsko above)
* It is possible to manage to trigger the animations with a single state name, while varying the input parameters continuously under the hood
* It is **critical** to understand when to mutate the animation state object and when to treat it as an immutable object:
   - _mutate_ the animation state object when changing an input parameter that _should not_ trigger the animation, but affect an otherwise newly triggered animation (in my data-driven examples, this applies to `duration` controlling the, no surprise, animation duration)
   - treat the animation state object as an _immutable_ object for changes that _should_ trigger the animation (in my examples, these are the changes related to `opacity`)
* As a result there is some lifting to be  done in the component lifecyle hooks (i.e. `ngOnChanges` to manage the animation state object(s)

I still have some investigation to do with respect to _exit_ animations apparently not being triggered.

Hope this helps as a starter illustration, suggestions welcome.

**Comment by manuelnaturalcycles** [comment]
i also think the `AnimationOptions` annotation is old.

`export interface AnimationOptions {
    delay?: number | string;
    params?: {
        [name: string]: any;
    };
}`

**Comment by MrCroft** [comment]
@tomwanzek @manuelnaturalcycles  I have an issue opened for the _**state() not being aware of input params**_ problem: #17362 - hope it pans out soon.

@tomwanzek - thank you for mentioning `AnimationBuilder` - wasn't even aware of it until now... So far, I was also using the _trick_ you mentioned (setting the desired end style properties on the animated element, which doesn't interfere with the animation, and then the properties are there when the animation ends).
But I'll check out `AnimationBuilder`, maybe it will help my case until #17362 will get some attention (if it will). Thanks!

**Comment by manuelnaturalcycles** [comment]
@tomwanzek any news regarding the `[style.opacity]="fadeStatus.params.end"` hack?

**Comment by azamanaza** [comment]
I have created an issue that is pretty much the same as mentioned in the earlier comments about end state not persisting.

https://github.com/angular/angular/issues/18423

We can close that one if it's decided to track this one instead.

**Comment by tomwanzek** [comment]
@manuelnaturalcycles I have not yet re-tested everything with Angular 4.3, but based on the comment by @azamanaza, it seems the issue of persisted end state has not yet been addressed.

Recognizing all the heroic efforts of the Angular Core Team 💯, I would have hoped for at least some interim indication from the Angular Core Team by now. Issue #17362 created by @MrCroft seems to be idle as well in this regard.

_It does not always have to be a fix right away, but some transparency to create predictability for the user community would be highly appreciated._

Alas, I have another animation issue (#16427 opened on April 28th with a fundamental question about support for SVG-related animations), which has not even been graced with an initial assessment/guidance despite several pings.

So, here is to the hope for some fresh insight. 🍻

**Comment by GustavoCostaW** [comment]
Any news about that? I really need expressions like this:

```
...
  animations: [
    trigger('state', [
      state('state1', style({
        background: `${this.background}`
      })),
      state('state2', style({
        background: `${this.background}`
      }))
    ])
  ]
...
```

**Comment by lazarljubenovic** [comment]
@GustavoCostaW What do you expect `this` to refer to in that snippet? Metadata is outside the scope of the class, so it simply cannot be an instance of the component.

**Comment by matsko** [comment]
@GustavoCostaW https://github.com/angular/angular/pull/18468

**Comment by GustavoCostaW** [comment]
@matsko great job! congratz bro, thank you :-)

**Comment by nop1984** [comment]
@matsko , super, thank you very much for this!
But i was not able to find any documentation about style expressions. 
@GustavoCostaW , did you managed to do background: `${this.background}` trick ?

**Comment by GustavoCostaW** [comment]
@nop1984 Check the link above that @matsko sent, and search the spec files to make possible dynamic styles.

**Comment by nop1984** [comment]
@GustavoCostaW thanks for answer. I was in his link and in my angular sources.
I was playing for 4 days with angular5 but no result
Also angular complains on syntax style('$value') (cant be sting, either '*' either object expected)

What i basically trying to do is simplified template of 2 not nested divs, where the width of first is a margin of second, animated.

**Comment by matsko** [comment]
Here's the plan for this in the upcoming 5.x releases: https://github.com/angular/angular/issues/20812

**Comment by GustavoCostaW** [comment]
@nop1984 thank you for the UPD =]

**Comment by jdmaguire** [comment]
😢

**Comment by thePunderWoman** [comment]
@jdmaguire We closed it because it's obsolete with the new animation APIs being added. Give those a try and let us know what you think.

**Comment by thePunderWoman** [comment]
This issue was closed as obsolete due to the introduction of the new `animate.enter` and `animate.leave` APIs. The `@angular/animations` package is now deprecated and will be removed in Angular v23. We hope you give the new APIs a try.
